solr/solrj/src/java/org/apache/solr/client/solrj/impl/ConcurrentUpdateJettySolrClient.java

```diff
@@ -0,0 +1,198 @@
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Locale;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import org.apache.solr.client.solrj.SolrRequest;
+import org.apache.solr.client.solrj.request.UpdateRequest;
+import org.apache.solr.client.solrj.util.ClientUtils;
+import org.apache.solr.common.params.ModifiableSolrParams;
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.common.params.UpdateParams;
+import org.eclipse.jetty.client.InputStreamResponseListener;
+import org.eclipse.jetty.client.OutputStreamRequestContent;
+import org.eclipse.jetty.client.Request;
+import org.eclipse.jetty.http.HttpMethod;
+
+/** A ConcurrentUpdate SolrClient using {@link Http2SolrClient}. */
+public class ConcurrentUpdateJettySolrClient extends ConcurrentUpdateHttp2SolrClient {
+  protected static final Charset FALLBACK_CHARSET = StandardCharsets.UTF_8;
+
+  private final Http2SolrClient client;
+
+  public static class Builder extends ConcurrentUpdateHttp2SolrClient.Builder {
+    /**
+     * @see ConcurrentUpdateHttp2SolrClient.Builder#Builder(String, HttpSolrClientBase)
+     */
+    public Builder(String baseUrl, Http2SolrClient client) {
+      this(baseUrl, client, false);
+    }
+
+    /**
+     * @see ConcurrentUpdateHttp2SolrClient.Builder#Builder(String, HttpSolrClientBase, boolean)
+     */
+    public Builder(String baseSolrUrl, Http2SolrClient client, boolean closeHttpClient) {
+      super(baseSolrUrl, client, closeHttpClient);
+      this.idleTimeoutMillis = client.getIdleTimeoutMillis();
+    }
+
+    @Override
+    public ConcurrentUpdateJettySolrClient build() {
+      return new ConcurrentUpdateJettySolrClient(this);
+    }
+  }
+
+  protected ConcurrentUpdateJettySolrClient(Builder builder) {
+    super(builder);
+    this.client = (Http2SolrClient) builder.client;
+  }
+
+  @Override
+  protected InputStreamResponseListener doSendUpdateStream(
+      ConcurrentUpdateHttp2SolrClient.Update update) throws IOException, InterruptedException {
+    InputStreamResponseListener responseListener;
+    try (OutStream out = initOutStream(basePath, update.request(), update.collection())) {
+      ConcurrentUpdateHttp2SolrClient.Update upd = update;
+      while (upd != null) {
+        UpdateRequest req = upd.request();
+        if (!out.belongToThisStream(req, upd.collection())) {
+          // Request has different params or destination core/collection, return to queue
+          queue.add(upd);
+          break;
+        }
+        send(out, upd.request(), upd.collection());
+        out.flush();
+
+        notifyQueueAndRunnersIfEmptyQueue();
+        upd = queue.poll(pollQueueTimeMillis, TimeUnit.MILLISECONDS);
+      }
+      responseListener = out.getResponseListener();
+    }
+    return responseListener;
+  }
+
+  private static class OutStream implements Closeable {
```

Comment:
From this point to the bottom, it's all moved from Http2SolrClient.  Since these methods are tightly related to concurrent updates, they seem more appropriate here, and were distracting where they were IMO.